src/lib/registration/rollback_script.rb

```diff
@@ -0,0 +1,72 @@
+# ------------------------------------------------------------------------------
+# Copyright (c) 2018 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify it under
+# the terms of version 2 of the GNU General Public License as published by the
+# Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+#
+# ------------------------------------------------------------------------------
+
+require "yast"
+require "fileutils"
+
+module Registration
+  # This class handles creating a registration rollback script  which is called
+  # when the upgrade is aborted or YaST crashes. The script body is stored
+  # in the data/registration/registration_rollback.sh file.
+  class RollbackScript
+    include Yast::Logger
+
+    BACKUP_DIR = "var/adm/backup/system-upgrade".freeze
+
+    # use number 0200, the original repositories need to be restored first,
+    # they are stored in script with number 0100
+    DEFAULT_SCRIPT_NAME = "restore-0200-registration.sh".freeze
+
+    SUSE_CONNECT = "/usr/sbin/SUSEConnect".freeze
+
+    attr_reader :name, :root
+
+    # @param name [String] script name
+    # @param root [String] target root
+    def initialize(name: DEFAULT_SCRIPT_NAME, root: "/mnt")
+      @name = name
+      @root = root
+    end
+
+    # create the registration rollback script
+    # @note The script can be created only when the target root is mounted.
+    def create
+      log.info "Creating registration rollback script #{script_path}"
+      src_file = File.expand_path("../../../data/registration/registration_rollback.sh", __FILE__)
```

Comment:
This looks wrong for me as it breaks y2dir, directory precedence and so on. It should look like
```
  Yast::Directory.find_data_file("registration/registration_rollback.sh")
```

Reply:
Um, I do not think it's a problem. And we already use this solution in several places/packages.

If you set a custom `Y2DIR` then also the `lib/registration/rollback_script.rb` file is read from this custom directory. And that data file is read from the path relative to the `rollback_script.rb` script. Which means it transitively honors the `Y2DIR` value.

Reply:
Ah, `y2update` ... This is the case which I forgot...

1) this approach is used elsewhere, we would need to fix more places
2) I considered it as a standard Ruby way (but it turned out that the YaST case is more complicated)